icefoganalytics/travel-authorization
# Rebuild the “Manager View” To Support Virtual “Supervisor” Role To Manage Requests Assigned to Them

Fixes https://github.com/icefoganalytics/travel-authorization/issues/90

# Context

See https://governmentofyukon.slack.com/archives/C04ETKKJS1X/p1699391682746329 for discussion.
> The intention for this round is we don't validate you have assigned the request to the correct person we anticipate you as an employee know the correct process.  So if you have been assigned as the approver and you don't have the proper authority you reassign it to someone who does.  Can we make the logic that you see anything that has been assigned to you.

See `web/src/modules/travel-authorizations/pages/ManagerView.vue` for current component. This needs to be rebuilt to support viewing by any user that has a "travel authorization" submitted to them, a.k.a. the "virtual" supervisor role.
Data model will likely need to be updated.

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/64e8c0a9-d46a-45aa-a81c-a46036d70001)

## Reflections

I might make sense to make the supervisor role real, and have it auto-granted to a user whenever they get a travel authorization request submitted to them?

## Additional Info
See "My Travel Requests" page for an example
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/779fb113-ef48-4a21-b18d-8dcb3d25538e)

Component is `web/src/modules/travel-authorizations/components/MyTravelAuthorizationsTable.vue`
At this point I need to break things up by user type, so might need a pre-work refactor.

I'll either go with:
- `/travel-authorization/<role>/SomeComponent`
- or `/<role>/travel-authorization/SomeComponent`

# Implementation

1. Decide on a standard naming convention of `<action/restriction><object><description><component-type>`
    e.g
    - `pages/EditMyTravelAuthorizationEstimatePage` - edit action for travel requests belonging to the current user (my) in regards to estimates, with Page type. Pages are things you can route to.
    - `pages/MyTravelAuthorizationsPage` - multi-action component for travel requests belonging to the current user
    - `pages/ManageTravelAuthorizationsPage` - manage (action) for all travel authorization accessible by the manage role
    - `layouts/EditMyTravelAuthorizationLayout` - layout for the EditMyTravelAuthorizationPage
    - `components/manage-travel-authorizations-page/ApprovedTripsCard` - component that is specific to the `ManageTravelAuthorizationsPage`.
2. Normalize email casing for all email fields on Sequelize models, and add setter normalize fields on create/update.
3. Fix policy scoping so that correct number of records are returned. Scoping now requires Sequelize scope/where clause logic. Users can now access `TravelAuthorizations` assigned to them via `supervisorEmail`
4. Build Manager View widgets as tables with per-status queries, and a `supervisorEmail` matching the currentUser email.
5. All Manager View tables now load data independently.
6. Debounce `SearchableUserEmailCombobox` so it doesn't hammer the back-end and the external provider.

---

Current widget to status mappings are:
1. Pending Approvals -> `submitted` status
2. Awaiting Expense Approval -> `expense_claim` status
3. Awaiting Changes -> `change_requested` status
4. Approved Trips -> `approved` status

They are _very_ easy to change, so tweaking them won't be a big deal later on.

# Screenshots

New - pretty much the same as the old, but table wiring is valid, and content gets returned
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/6136bbf8-4695-40e9-84f4-d53abf6e22e6)

# Testing Instructions

1. Boot the app via `dev up`.
8. Log in to the app at http://localhost:8080.
9. Go to the "My Travel Requests" page via the top nav dropdown.
10. Create a new travel authorization via the "+ Travel Authorization" button.
11. Fill in all the fields, and submit it to you, as the supervisor.
12. Go to the "Manager View" via the top nav dropdown.
13. Check that the request shows up in the "Pending Approvals" table widget.
    > Note: you can't actually approve anything yet, just see the data.
